File: PageObjectDotNet/AutomationPractice/PageObjects/AutomationPracticeStore/Objects/ProductInCartPopUp.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using SeleniumFramework.Base;
using System;
using System.Collections.Generic;

namespace AutomationPractice.PageObjects.AutomationPracticeStore.Objects
{
    public class ProductInCartPopUp : BasePage
    {
        private readonly By labelProductName = By.Id("layer_cart_product_title");
        private readonly By labelProductAttributes = By.Id("layer_cart_product_attributes");
        private readonly By labelProductQuantity = By.Id("layer_cart_product_quantity");
        private readonly By labelProductTotalPrice = By.Id("layer_cart_product_price");

        private readonly By buttonContinueShopping = By.ClassName("continue");
        private readonly By buttonProceedCheckout = By.CssSelector("a.btn-default");
        private readonly By buttonClosePopup = By.ClassName("cross");
        private readonly By labelNumberOfItemsInCart = By.ClassName("ajax_cart_quantity");
        private readonly By labelTotalProductsAmount = By.ClassName("ajax_block_products_total");
        private readonly By labelTotalShippingAmount = By.CssSelector("div.layer_cart_row span.ajax_cart_shipping_cost");
        private readonly By labelTotalFinalAmount = By.CssSelector("div.layer_cart_row span.ajax_block_cart_total");

        private readonly By popUpLocator = By.CssSelector("div#layer_cart div.clearfix");

        private readonly string PLURAL_PHRASE_LOCATOR = "ajax_cart_product_txt_s";
        private readonly string SINGULAR_PHRASE_LOCATOR = "ajax_cart_product_txt";

        private readonly IWebElement popUp;
        private readonly WebDriverWait wait;


        public ProductInCartPopUp() : base() 
        {
            wait = new WebDriverWait(Driver, TimeSpan.FromSeconds(4));
            wait.Until(ExpectedConditions.ElementIsVisible(popUpLocator));
            popUp = Driver.FindElement(popUpLocator);
        }

        public string GetProductName() 
        {
            return popUp.FindElement(labelProductName).Text;
        }

        public int GetProductQuantity() 
        {
            return int.Parse(popUp.FindElement(labelProductQuantity).Text);
        }

        public double GetProductTotal() 
        {
            return double.Parse(popUp.FindElement(labelProductTotalPrice).Text.Substring(1));
        }

        public IReadOnlyList<string> GetProductAttributes() 
        {
            var attrs = popUp.FindElement(labelProductAttributes).Text.Replace(" ", string.Empty);
            return attrs.Split(',');
        }

        public double GetTotalProductsPrice() 
        {
            return double.Parse(popUp.FindElement(labelTotalProductsAmount).Text.Substring(1));
        }

        public double GetTotalShippingCost() 
        {
            return double.Parse(popUp.FindElement(labelTotalShippingAmount).Text.Substring(1));
        }

        public double GetCartFinalAmount() 
        {
            return double.Parse(popUp.FindElement(labelTotalFinalAmount).Text.Substring(1));
        }

        public string GetTotalItemsPhrase() 
        {
            var classLocator = GetTotalItemsInCart() > 1 ? PLURAL_PHRASE_LOCATOR : SINGULAR_PHRASE_LOCATOR;
            return popUp.FindElement(By.ClassName(classLocator)).Text;
        }

        public int GetTotalItemsInCart() 
        {
            return int.Parse(popUp.FindElement(labelNumberOfItemsInCart).Text);
        }

        public CheckoutPage ProceedToCheckout() 
        {
            popUp.FindElement(buttonProceedCheckout).Click();
            return new CheckoutPage();
        }

        public void ContinueShopping() 
        {
            popUp.FindElement(buttonContinueShopping).Click();
            wait.Until(ExpectedConditions.InvisibilityOfElementLocated(popUpLocator));
        }

        public void ClosePopup() 
        {
            popUp.FindElement(buttonClosePopup).Click();
            wait.Until(ExpectedConditions.InvisibilityOfElementLocated(popUpLocator));
        }

    }
}

File: PageObjectPython/PageObjects/AutomationPracticeStore/Objects/ProductInCartPopup.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support.expected_conditions import visibility_of_element_located
from selenium.webdriver.support.expected_conditions import invisibility_of_element_located

from PageObjects.AutomationPracticeStore.CheckoutPage import CheckoutPage
from PageObjects.BasePageObject import BasePageObject


class ProductInCartPopup(BasePageObject):

    __label_product_name = (By.ID, 'layer_cart_product_title')
    __label_product_attributes = (By.ID, 'layer_cart_product_attributes')
    __label_product_quantity = (By.ID, 'layer_cart_product_quantity')
    __label_product_total_price = (By.ID, 'layer_cart_product_price')

    __button_continue_shopping = (By.CLASS_NAME, 'continue')
    __button_proceed_checkout = (By.CSS_SELECTOR, 'a.btn-default')
    __button_close_popup = (By.CLASS_NAME, 'cross')
    __number_of_items_in_cart = (By.CLASS_NAME, 'ajax_cart_quantity')
    __label_total_products_amount = (By.CLASS_NAME, 'ajax_block_products_total')
    __label_total_shipping_amount = (By.CSS_SELECTOR, 'div.layer_cart_row span.ajax_cart_shipping_cost')
    __label_total_final_amount = (By.CSS_SELECTOR, 'div.layer_cart_row span.ajax_block_cart_total')

    __self_locator = (By.CSS_SELECTOR, 'div#layer_cart div.clearfix')

    __PLURAL_PHRASE_LOCATOR = "ajax_cart_product_txt_s"
    __SINGULAR_PHRASE_LOCATOR = "ajax_cart_product_txt"

    def __init__(self):
        self.__wait = WebDriverWait(self.base_page_driver, 4)
        self.__wait.until(visibility_of_element_located(self.__self_locator))
        self.__we = self.base_page_driver.find_element(*self.__self_locator)

    def get_product_name(self):
        return self.__we.find_element(*self.__label_product_name).text

    def get_product_quantity(self):
        return self.__we.find_element(*self.__label_product_quantity).text

    def get_product_total(self):
        return self.__we.find_element(*self.__label_product_total_price).text[1:]

    def get_product_attributes(self):
        attrs = str(self.__we.find_element(*self.__label_product_attributes).text).replace(' ', '')
        return attrs.split(',')

    def get_total_products_price(self):
        return self.__we.find_element(*self.__label_total_products_amount).text[1:]

    def get_total_shipping_cost(self):
        return self.__we.find_element(*self.__label_total_shipping_amount).text[1:]

    def get_cart_final_amount(self):
        return self.__we.find_element(*self.__label_total_final_amount).text[1:]

    def get_total_items_phrase(self):
        locator = self.__PLURAL_PHRASE_LOCATOR if self.get_total_items_in_cart() > 1 else self.__SINGULAR_PHRASE_LOCATOR
        return self.__we.find_element_by_class_name(locator).text

    def get_total_items_in_cart(self):
        return int(self.base_page_driver.find_element(*self.__number_of_items_in_cart).text)

    def proceed_to_checkout(self):
        self.__we.find_element(*self.__button_proceed_checkout).click()
        return CheckoutPage()

    def continue_shopping(self):
        self.__we.find_element(*self.__button_continue_shopping).click()
        self.__wait.until(invisibility_of_element_located(self.__self_locator))

    def close_popup(self):
        self.__we.find_element(*self.__button_close_popup).click()
        self.__wait.until(invisibility_of_element_located(self.__self_locator))
